### ISSUE TITLE
[ACF 6.0 RC1] Extra field settings are rendered twice when changing the field type

### ISSUE THREAD

**Issue by strategio** [AdvancedCustomFields/acf#705]
In the last stable version `5.12.3`, we need the following code to add an extra setting to a field:

```php
add_action( 'init', function() {
	$renderFooSetting = function( $field ) {
		acf_render_field_setting(
			$field,
			[
				'label'         => __( 'Foo', 'foo' ),
				'instructions'  => __( 'Foo description', 'foo' ),
				'type'          => 'radio',
				'name'          => 'foo',
				'layout'        => 'horizontal',
				'choices'       => [
					1 => __( 'One', 'foo' ),
					2 => __( 'Two', 'foo' ),
				],
				'default_value' => 1,
			]
		);
	};

	add_action( 'acf/render_field_settings', $renderFooSetting );

	$field = acf_maybe_get_POST( 'field' );
	if ( isset( $field['type'] ) ) {
		add_action( "acf/render_field_settings/type={$field['type']}", $renderFooSetting );
	}
} );
```

[acf-5.12.3-original-snippet.webm](https://user-images.githubusercontent.com/1941656/187695124-8b2eabc7-f619-4f8e-b33c-2e90e0616d71.webm)

Now if we omit the 2nd action on `acf/render_field_settings/type={$field['type']}` (it can just be commented out in the original snippet), the extra setting is not rendered when the field type is changed.

[acf-5.12.3-degraded-snippet.webm](https://user-images.githubusercontent.com/1941656/187695044-d7cc6cc3-829d-40a7-aaef-651c3a6d8e8c.webm)

**=> This is to show you that we currently need to hook on the 2 actions if we want a proper rendering of the extra setting in all contexts.**

Now with ACF 6.0 RC1 and the same original snippet, when we change the field type, the extra setting is rendered twice:

[acf-6.0-rc1-original-snippet.webm](https://user-images.githubusercontent.com/1941656/187695234-b4d3d3f2-9abd-40cc-93fd-603178789c77.webm)

We could easily patch our addon for ACF 6.0 to avoid hooking on `acf/render_field_settings/type={$field['type']}`. But it basically means that ACF 6.0 is not really backward compatible. If other 3rd party plugins complains about it, you will fix that and then my patched addon will be broken...

Could you please let me know if you are willing to fix this (make it backward compatible) or if we should fix it on our side? Thanks!

**Comment by lgladdy** [comment]
@strategio Hey Pierre, Could I just confirm the intended behaviour here is that the field setting should be shown on every field type?

My gut feeling here is we've actually fixed a bug we didn't even know existed, and this might be a case of us needing to just note this to other third party devs in RC2 and the final release notes.

**Comment by mattgrshaw** [comment]
Hi @strategio,

Thanks for raising this, it's great to get this sort of feedback before the official release. I think there's a couple issues at play here.

In your code sample above, the field setting is essentially getting registered twice. In ACF 5, the [recommended](https://www.advancedcustomfields.com/resources/adding-custom-settings-fields/) method for registering a setting for all field types would be to pass `true` as the third `$global` parameter to `acf_render_field_setting()`, for example:

```
add_action( 'init', function() {
	$renderFooSetting = function( $field ) {
		acf_render_field_setting(
			$field,
			[
				'label'         => __( 'Foo', 'foo' ),
				'instructions'  => __( 'Foo description', 'foo' ),
				'type'          => 'radio',
				'name'          => 'foo',
				'layout'        => 'horizontal',
				'choices'       => [
					1 => __( 'One', 'foo' ),
					2 => __( 'Two', 'foo' ),
				],
				'default_value' => 1,
			],
			true
		);
	};

	add_action( 'acf/render_field_settings', $renderFooSetting );
} );
```
I've tested this out on ACF 5 (and 6) and confirmed that the new setting gets added, even when changing field types, so you wouldn't need to add the second action that uses the field type in the `POST` request.

---

The default/expected behavior in both ACF 5 and 6 is to remove all non-global field settings whenever the field type is changed, so that the new field type settings can be loaded in.

But there is a bug on our end in ACF 6 where field types that are registered through the `acf/render_field_settings` action hook that aren't global weren't getting removed. So when the second `acf/render_field_settings/type={$field['type']}` action hook was getting fired, the setting was getting loaded in a second time and the first setting wasn't getting removed.

---

In terms of moving forward, I think you should update your code to ditch the second action hook and instead use the third `$global` parameter in `acf_render_field_setting()`. That should work in both ACF 5 and the current version of ACF 6.

We'll look into the bug on our end and make sure that we don't disrupt that functionality if we decide to address that for ACF 6.0.

**Comment by strategio** [comment]
Thanks for the feedback @mattgrshaw!

I ran some manual tests with the `$global` parameter set to `true` and it seems to work.
Our end-to-end test will confirm that and we'll include that fix in our next ACFML release (1.10.4).

**Comment by strategio** [comment]
@mattgrshaw, just to inform you that we released ACFML 1.10.4 today including a fix for this issue (as commented above).

The problem is closed on our side.

Now I don't know if you want to investigate further, so I'll let you close the ticket if you wish.

Thanks again :)

**Comment by mattgrshaw** [comment]
Awesome, glad to hear this is sorted on your end!

I think we'll close this on our end as well (barring any similar reports) as it can only happen when a field is registered on the global action hook but not explicitly marked as global, which would also be broken in ACF 5 if not for that second action hook. 

Please let us know if you notice any further issues with the RC or otherwise.
